### PR TITLE
Improve wallet rescan API

### PR DESCRIPTION
**Pull request by ryanofsky**
This PR is just refactoring. There are no changes in behavior. It implements a request from @TheBlueMatt in https://github.com/bitcoin/bitcoin/pull/9827#issuecomment-282123202 to make rescan calls use timestamps instead of block index pointers.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
utACK 4182e6c89640f8745f0461f5ef610f343ef26296

**Comment by ryanofsky** [comment]
Rebased 2de5c79de7536c0edbfa4e0cb0175a9b8608d720 -> 9bb66ab66013fc3024fad00bceeb22a330b5bc1b ([pr/scanclean.2](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.2) -> [pr/scanclean.3](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.3))

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1459,6 +1463,34 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
 }
 
 /**
+ * Scan active chain for relevant transactions after importing keys. This should
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: Shouldnt the docs (here and above) be in the .h and not the .cpp?

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -1459,6 +1463,34 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
...
+ * be called whenever new keys are added to the wallet, with the oldest key
+ * creation time minus TIMESTAMP_WINDOW.
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: Maybe move TIMESTAMP_WINDOW into an implicit thing that this function applies to the provided timestamp (and move its definition out of wallet.h)?

**Review by TheBlueMatt — COMMENTED** [review]
utACK 9bb66ab66013fc3024fad00bceeb22a330b5bc1b

**Review by ryanofsky — COMMENTED** [review]
Added 2 commits 9bb66ab66013fc3024fad00bceeb22a330b5bc1b -> deaf48b046e573f6774d19e74b58918ed777cb14 ([pr/scanclean.3](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.3) -> [pr/scanclean.5](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/scanclean.3...pr/scanclean.5))

**Comment by laanwj** [comment]
utACK deaf48b.

**Pull request merged**